### ISSUE TITLE
ASR Bug: `ExplicitDeallocate.vars" is not a `symbol *`

### ISSUE THREAD

**Issue by rebcabin** [lcompilers/lpython#1492]
This tripped up my abstract-evaluator:

The ASR spec says:

```clojure
;; | ExplicitDeallocate(symbol* vars)
```

but the reality is that `vars` is a vector of even length: `[<stid> <symbol>, ...]`, where `stid` is an integer global symbol-table id:

<img width="625" alt="image" src="https://user-images.githubusercontent.com/207047/216429113-d459413d-9a90-441b-9484-96b91b3e1c8e.png">

**Comment by certik** [comment]
That is how `symbol` is printed in the Clojure-like output.

Similar issue to https://github.com/lcompilers/lpython/issues/1478

**Comment by certik** [comment]
A symbol (reference) could be printed as `(2 a)` instead of just `2 a`.

How should `Var` be printed? It could be `Var((2 a))`, instead of the current `2 a`. There is some inconsistency currently.

**Comment by rebcabin** [comment]
I strongly disagree with `symbol` meaning different things depending on context, say a `Function` or `Module`, here, and an implicitly paired integer `id` and an identifier elsewhere. This complicates interpretation and test generation. My software must dispatch not only the head of symbol (like `Function`, `Module`, etc.) but on non-robust structural or implicit contextual cues.

I recommend that `2 x` be (first) explicitly tupled, as in `(2 x)`, and (second) be added to the `tuple` group as a `symbol_ref` perhaps like this:

```
symbol_ref = (symbol_table_id stid, identifier symnym)
```

Statements like `ExplicitDeallocate` that take arguments like `[2 x 2 y]` should be specified to take one or more (a vector of) `symbol_ref+` or `symbol_ref*` (instead of `symbol*`) and take and print arguments like `[(2 x) (2 y)]`. 

Emphatically, again, reading and writing `symbol`s differently depending on context is a slippery slope: we should not give ourselves permission for ambiguity and looseness.

## Details

I have a partially automated recursive-descent evaluator. To it, `symbol` is, as written in ASR.asdl, one of `Program`, `Module`, `Function`, etc. It recursively evaluates parameters based on their terms (`symbol`, `stmt`, `expr`, etc.) These methods can be automatically generated from the ASDL, but _only_ if `symbol` means `symbol` and nothing else! :

```clojure
(defmulti eval-symbol first)

;; symbol  -- QUOTED from ASR.asdl
;;     = Program(symbol_table symtab, identifier name,
;;         identifier* dependencies, stmt* body)

(defmethod eval-symbol 'Program
  [[head    symtab    nym    dependencies    body
    :as program]]
  (fn [penv]
    {... 
     :symtab       ((eval-symbol symtab)       penv)   ; 'SymbolTable
     :nym          ((eval-node   nym)          penv)   ; identifier
     :dependencies ((eval-node   dependencies) penv)   ; identifier *
     :body         ((eval-stmts  body) penv)           ; stmt *
     :penv         penv
     }))

(defmethod eval-symbol 'SymbolTable
  [[head    integer-id    bindings
    :as symbol-table]]
  (fn [penv]
etc. ...
  ))

;; | Variable(symbol_table parent_symtab, identifier name,
;;   identifier* dependencies, intent intent,
;;   expr? symbolic_value, expr? value,
;;   storage_type storage, ttype type,
;;   abi abi, access access, presence presence, bool value_attr)

(defmethod eval-symbol 'Variable
  [[head     parent-symtab-id  nym     dependencies
    intent   symbolic-value    value   storage
    type-    abi               access  presence  value-attr 
    :as variable]]
  (fn [penv]
etc. ...
  ))

(defmethod eval-symbol 'Function
  [[head                       ; 'Function
  (fn [penv]
etc. ...
  ))
```

When another term says `symbol`, `symbol?` or `symbol*`, I expect to call `(eval_symbol <foo>)` appropriately. To work around this bug, `(eval-symbol ...)` cannot simply dispatch on the head, as above, but must non-robustly figure out from the structure whether the thing being passed in is actually a symbol or a symbol-ref.

**Comment by rebcabin** [comment]
I'll get to critique of `Var` later. I'm building the evaluator piece-by-piece and haven't come to `Var` yet.

**Comment by rebcabin** [comment]
Editing my comment above ...

**Comment by rebcabin** [comment]
I'm still debating the possibilities with myself ...

if `SymbolRef` is an `expr`, then its appearance in other ASR terms will be of type `expr` or `expr*` etc. That's less than ideal, because there is a hidden semantical restriction that the expr must be a `SymbolRef`.

If `symbol_ref` is a headless tuple, as in `(symtab_id id, identifier symnym)`, then that begs unification with `Var`, leading to a clumsy print like `( Var (2 x) )`.

Perhaps the best solution is  for a new, top-level, singleton composite:

```
symref = SymbolRef(symtab_id id, identifier symnym)
```

Later, if not overengineering, it could grow alternatives:

```
symref = VariableRef(symtab_id id, identifier varnym)
       | FunctionRef(symtab_id, identifier funnym)
       | ModuleRef(symtab_id, identifier modnym)
... -- one for every symbol
```

but that feels like overengineering because it repeats `symbol` -- it's not DRY.

**Comment by czgdp1807** [comment]
Note: I changed `ExplicitDeallocate.vars` to `expr` (https://github.com/lfortran/lfortran/pull/1188/commits/03e02742f73d0db7deb8bb0006de8c6a685f0a0a#diff-65d4f5aefdce4101f9c116614e2f04697eeb987fa16565c21070c0daf785c4e5) because it is required to compile FPM with LFortran. It would be applied to LPython soon. Thanks.
